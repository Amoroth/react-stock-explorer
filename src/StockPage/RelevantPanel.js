import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'

import styles from './StockPage.module.css'

const relevantPanel = ({ name, change, symbol, close, onSelect }) => {
  let shortCompanyName = name
  const priceChange = `${change > 0 ? '+' : ''}${(
    change * 100
  ).toFixed(2)}%`

  if (shortCompanyName.length > 14) {
    shortCompanyName = `${shortCompanyName.slice(0, 12)}...`
  }

  return (
    <div className={styles['relavent-item']} key={symbol}>
      <div
        className={`${styles['relavent-item-row']} ${
          styles['relavent-item-title']
        }`}
      >
        <h6
          title={
            shortCompanyName[shortCompanyName.length - 2] === '.'
              ? name
              : null
          }
        >
          {shortCompanyName}
        </h6>
        <span>{symbol}</span>
      </div>
      <div
        className={`${styles['relavent-item-row']} ${
          styles['relavent-item-info']
        }`}
      >
        <span>{close}</span>
        <span
          style={{ color: change > 0 ? 'green' : 'red' }}
        >
          {priceChange}
        </span>
        <Link
          to={`/stock?cmp=${symbol}`}
          onDragStart={(e) => e.preventDefault()}
          onClick={() => onSelect(symbol)}
        >
          More...
        </Link>
      </div>
    </div>
  )
}

relevantPanel.propTypes = {
  name: PropTypes.string.isRequired,
  change: PropTypes.number.isRequired,
  symbol: PropTypes.string.isRequired,
  close: PropTypes.number.isRequired,
  onSelect: PropTypes.func.isRequired,
}

export default relevantPanel
