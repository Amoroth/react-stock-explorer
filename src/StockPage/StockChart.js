import React from 'react'
import { VictoryLine, VictoryChart, VictoryAxis, VictoryTheme } from 'victory'

const stockChart = ({ data }) => {
  const graphSizeX = window.innerWidth < 620 ? 350 : 500
  const graphSizeY = window.innerWidth < 620 ? 400 : 250

  return (
    <svg
      role="img"
      style={{
        position: 'static',
        pointerEvents: 'none',
        width: '100%',
        height: '100%',
      }}
      viewBox={`0 0 ${graphSizeX} ${graphSizeY}`}
    >
      <VictoryChart
        height={graphSizeY}
        width={graphSizeX}
        theme={VictoryTheme.material}
        standalone={false}
        style={{ parent: { position: 'static' } }}
        domainPadding={{ y: [20, 20] }}
        padding={{ top: 10, right: 40, bottom: 30, left: 50 }}
      >
        <VictoryAxis
          tickValues={[1, 2, 3, 4]}
          tickFormat={(y) => `${y}`}
          fixLabelOverlap
          style={{
            axis: { stroke: '#344955' },
            tickLabels: { fill: '#344955' },
          }}
        />
        <VictoryAxis
          dependentAxis
          tickFormat={(x) => `${x}`}
          style={{
            axis: { stroke: '#344955' },
            tickLabels: { fill: '#344955' },
          }}
        />
        <VictoryLine
          data={data}
          x="label"
          y="close"
          style={{ data: { stroke: '#344955' }, labels: { display: 'none' } }}
        />
      </VictoryChart>
    </svg>
  )
}

export default stockChart
