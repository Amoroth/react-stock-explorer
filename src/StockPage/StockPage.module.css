.container {
  margin: auto;
  width: 60%;
  max-width: 800px;
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 2px 4px 10px rgba(0,0,0,.025);
  padding: 16px;
}