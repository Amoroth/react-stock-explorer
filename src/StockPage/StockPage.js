import React, { Component } from 'react'
import Draggable from 'react-draggable'
import { Link } from 'react-router-dom'

import PageTitle from '../shared/PageTitle'
import StockChart from './StockChart'
import Spinner from '../shared/Spinner'
import styles from './StockPage.module.css'

class StockPage extends Component {
  state = {
    book: {
      symbol: '',
      companyName: '',
      primaryExchange: '',
      sector: '',
      open: 0,
      openTime: 0,
      close: 0,
      closeTime: 0,
      high: 0,
      low: 0,
      previousClose: 0,
      change: 0,
      week52High: 0,
      week52Low: 0
    },
    chart: [],
    chartTime: 0,
    relevant: [],
    logo: ''
  }

  componentDidMount() {
    const cmp = new URLSearchParams(this.props.location.search.slice(1)).get('cmp')
    this.updateBook(cmp)
    this.updateChart(cmp, '1m')
  }

  componentDidUpdate() {
    const cmp = new URLSearchParams(this.props.location.search.slice(1)).get('cmp')
    if (cmp.toLowerCase() !== this.state.book.symbol.toLowerCase()) {
      this.updateBook(cmp)
      this.updateChart(cmp, '1m')
    }
  }

  updateBook = (cmp) => {
    fetch(`https://api.iextrading.com/1.0/stock/${cmp}/batch?types=quote,relevant,logo`)
      .then((res) => {
        return res.json()
      }).then((json) => {
        fetch(`https://api.iextrading.com/1.0/stock/market/batch?types=quote&symbols=${json.relevant.symbols.join(',')}&filter=companyName,symbol,changePercent,close`)
          .then((response) => {
            return response.json()
          }).then((endJson) => {
            let relevantCompanies = []
            for (let key in endJson) {
              let tempCompany = {...endJson[key].quote}
              tempCompany.symbol = key
              relevantCompanies.push(tempCompany)
            }
            this.setState({
              book: json.quote,
              relevant: relevantCompanies,
              logo: json.logo.url
            })
          })
      })
  }

  updateChart = (symbol, time) => {
    fetch(`https://api.iextrading.com/1.0/stock/${symbol}/chart/${time}?filter=date,close,label`).then((res) => {
      return res.json()
    }).then((json) => {
      let timeNum = ['1m', '3m', '6m', '1y', '2y'].indexOf(time)
      this.setState({
        chart: json,
        chartTime: timeNum
      })
    })
  }

  onSelectCharttime = (time) => {
    this.setState({
      chart: []
    })
    this.updateChart(this.state.book.symbol, time)
  }

  render() {
    let chartEl = (
      <div className={styles['spinner-container']}>
        <Spinner />
      </div>
    )
    if (this.state.chart.length > 0) {
      chartEl = <StockChart data={this.state.chart} />
    }

    return (
      (
        <div className={styles['container']}>
          <PageTitle
            logo={this.state.logo}
            name={this.state.book.companyName}
            exchange={this.state.book.primaryExchange}
            symbol={this.state.book.symbol}
            link
          />
          <hr />
          <div>
            <div className={styles['charttime-container']}>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 0 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('1m')}>1m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 1 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('3m')}>3m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 2 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('6m')}>6m</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 3 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('1y')}>1y</button>
              <button className={`${styles['charttime-button']} ${this.state.chartTime === 4 ? styles['charttime-button-active'] : null}`} onClick={() => this.onSelectCharttime('2y')}>2y</button>
            </div>
            { chartEl }
            <div className={styles['book-info']}>
                <span>Open:</span> <span>{this.state.book.open} USD</span>
                <span>Prev. Close:</span> <span>{this.state.book.previousClose} USD</span>
                <span>Max:</span> <span>{this.state.book.high} USD</span>
                <span>Highest/52 weeks:</span> <span>{this.state.book.week52High} USD</span>
                <span>Min:</span> <span>{this.state.book.low} USD</span>
                <span>Lowest/52 weeks</span> <span>{this.state.book.week52Low} USD</span>
                <span>Market Cap</span> <span>{(this.state.book.marketCap / 1000000000).toFixed(2)} mld USD</span>
                <span>P/E Ratio</span> <span>{this.state.book.peRatio} USD</span>
            </div>
          </div>
          <hr />
          <div>
            {this.state.book.sector ? <p>Sector: {this.state.book.sector}</p> : <p>Relevant:</p>}
            <div className={styles['relevant']}>
              <Draggable
                axis="x"
                defaultPosition={{x: 0, y: 0}}
                bounds={{left: -1010, right: 0}}
              >
                <div className={styles['relevant-inner']}>
                  {this.state.relevant.map((val) => {
                    let shortCompanyName = val.companyName
                    let priceChange = `${val.changePercent > 0 ? '+' : ''}${(val.changePercent * 100).toFixed(2)}%`

                    if (shortCompanyName.length > 14)
                      shortCompanyName = shortCompanyName.slice(0, 12) + '...'
                    return (
                      <div className={styles['relavent-item']} key={val.symbol}>
                        <div className={`${styles['relavent-item-row']} ${styles['relavent-item-title']}`}>
                          <h6 title={ shortCompanyName[shortCompanyName.length - 2] === '.' ? val.companyName : null }>{shortCompanyName}</h6>
                          <span>{val.symbol}</span>
                        </div>
                        <div className={`${styles['relavent-item-row']} ${styles['relavent-item-info']}`}>
                          <span>{val.close}</span>
                          <span style={{color: val.changePercent > 0 ? 'green' : 'red'}}>
                            { priceChange }
                          </span>
                          <Link
                            to={`/stock?cmp=${val.symbol}`}
                            onDragStart={(e) => e.preventDefault()}
                            onClick={() => window.scrollTo(0, 0)}
                          >More...</Link>
                        </div>
                      </div>
                    )
                  })}
                </div>
              </Draggable>
            </div>
          </div>
          <hr />
          <div>
            <p>Informacje Finansowe (osobne okno?)</p>
            <p>O firmie (osobne okno?)</p>
          </div>
        </div>
      )
    )
  }
}

export default StockPage