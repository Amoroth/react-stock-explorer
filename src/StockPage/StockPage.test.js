import React from 'react'
import { shallow } from 'enzyme'

import StockPage from './StockPage'

describe('<StockPage />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      location: { search: '&cmp=COM' },
      favorites: [],
      onFavorite: jest.fn(),
      currency: 'USD',
      currencyFormat: jest.fn((e) => e),
    }
    wrapper = shallow(<StockPage {...props} />)

    wrapper.setState({
      chart: [
        { label: 'a', close: 1 },
        { label: 'b', close: 2 },
        { label: 'c', close: 3 },
      ],
    })
  })

  it('renders', () => {
    wrapper.setState({
      relevant: [
        { companyName: 'Company', symbol: 'COM', close: '2', changePercent: 0.02 },
        { companyName: 'Other', symbol: 'OTH', close: '3', changePercent: 0.03 },
        { companyName: 'Other 2', symbol: 'OT2', close: '4', changePercent: 0.04 },
      ],
    })

    expect(wrapper).toMatchSnapshot()
  })

  it('shows spinner if no chart data', () => {
    wrapper.setState({ chart: [] })
    expect(wrapper.find('.spinner-container').exists()).toBe(true)
    expect(wrapper.find('stockChart').exists()).toBe(false)
  })

  it('shows chart if chart data exists', () => {
    expect(wrapper.find('.spinner-container').exists()).toBe(false)
    expect(wrapper.find('stockChart').exists()).toBe(true)
  })

  it('shows p with sector if sector exists', () => {
    wrapper.setState({ book: { sector: 'Supreme' } })

    expect(wrapper.find('p').text()).toBe('Sector: Supreme')
  })

  it('shows p with relevant if no sector', () => {
    expect(wrapper.find('p').text()).toBe('Relevant:')
  })

  it('shows relevant items', () => {
    expect(wrapper.find('relevantPanel').length).toBe(0)

    wrapper.setState({
      relevant: [
        { companyName: 'Company', symbol: 'COM', close: '2', changePercent: 0.02 },
        { companyName: 'Other', symbol: 'OTH', close: '3', changePercent: 0.03 },
        { companyName: 'Other 2', symbol: 'OT2', close: '4', changePercent: 0.04 },
      ],
    })

    expect(wrapper.find('relevantPanel').length).toBe(3)
  })

  it('renders all book info', () => {
    expect(wrapper.find('.book-info span').length).toBe(20)
  })

  it('formats book info item', () => {
    wrapper.setState({ book: { close: 2, latestPrice: 3 } })

    expect(wrapper.find('.book-info span').at(0).text()).toBe('Close:')
    expect(wrapper.find('.book-info span').at(1).text()).toBe('2 USD')
    expect(wrapper.find('.book-info span').at(2).text()).toBe('Latest Price:')
    expect(wrapper.find('.book-info span').at(3).text()).toBe('3 USD')
  })

  it('changes chart time', () => {
    const mockFn = jest.fn()
    wrapper.instance().onSelectCharttime = mockFn

    wrapper.find('.charttime-button').last().simulate('click')
    expect(mockFn).toBeCalledWith('2y')
  })
})
