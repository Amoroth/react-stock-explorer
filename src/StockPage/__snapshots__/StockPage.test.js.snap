// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockPage /> renders 1`] = `
<div
  className="container"
>
  <withRouter(pageTitle)
    exchange=""
    favorite={false}
    link={true}
    logo=""
    name=""
    onFav={[MockFunction]}
    symbol=""
  />
  <hr />
  <div>
    <div
      className="charttime-container"
    >
      <button
        className="charttime-button charttime-button-active"
        key="1m"
        onClick={[Function]}
        type="button"
      >
        1m
      </button>
      <button
        className="charttime-button null"
        key="3m"
        onClick={[Function]}
        type="button"
      >
        3m
      </button>
      <button
        className="charttime-button null"
        key="6m"
        onClick={[Function]}
        type="button"
      >
        6m
      </button>
      <button
        className="charttime-button null"
        key="1y"
        onClick={[Function]}
        type="button"
      >
        1y
      </button>
      <button
        className="charttime-button null"
        key="2y"
        onClick={[Function]}
        type="button"
      >
        2y
      </button>
    </div>
    <stockChart
      data={
        Array [
          Object {
            "close": 1,
            "label": "a",
          },
          Object {
            "close": 2,
            "label": "b",
          },
          Object {
            "close": 3,
            "label": "c",
          },
        ]
      }
    />
    <div
      className="book-info"
    >
      <span>
        Close:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Latest Price:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Change Percent:
      </span>
      <span>
        NaN%
      </span>
      <span>
        Change:
      </span>
      <span>
        0 USD
      </span>
      <span>
        High:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Week52 High:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Low:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Week52 Low:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Market Cap:
      </span>
      <span>
        0 USD
      </span>
      <span>
        Pe Ratio:
      </span>
      <span>
        0 USD
      </span>
    </div>
  </div>
  <hr />
  <div>
    <p>
      Relevant:
    </p>
    <div
      className="relevant"
    >
      <Draggable
        allowAnyClick={false}
        axis="x"
        bounds={
          Object {
            "left": 105.6,
            "right": 0,
          }
        }
        cancel={null}
        defaultClassName="react-draggable"
        defaultClassNameDragged="react-draggable-dragged"
        defaultClassNameDragging="react-draggable-dragging"
        defaultPosition={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        disabled={false}
        enableUserSelectHack={true}
        grid={null}
        handle={null}
        offsetParent={null}
        onDrag={[Function]}
        onMouseDown={[Function]}
        onStart={[Function]}
        onStop={[Function]}
        position={null}
        transform={null}
      >
        <div
          className="relevant-inner"
        >
          <relevantPanel
            change={0.02}
            close="2"
            key="rel-COM"
            name="Company"
            onSelect={[Function]}
            symbol="COM"
          />
          <relevantPanel
            change={0.03}
            close="3"
            key="rel-OTH"
            name="Other"
            onSelect={[Function]}
            symbol="OTH"
          />
          <relevantPanel
            change={0.04}
            close="4"
            key="rel-OT2"
            name="Other 2"
            onSelect={[Function]}
            symbol="OT2"
          />
        </div>
      </Draggable>
    </div>
  </div>
</div>
`;
